mindsdb/interfaces/agents/langchain_agent.py

```diff
@@ -0,0 +1,325 @@
+from langchain.agents.initialize import initialize_agent
+from langchain.chains.conversation.memory import ConversationSummaryBufferMemory
+from langchain.schema import SystemMessage
+from langchain_community.chat_models import ChatAnthropic, ChatOpenAI, ChatAnyscale, ChatLiteLLM, ChatOllama
+from langchain_core.prompts import PromptTemplate
+from langfuse import Langfuse
+from langfuse.callback import CallbackHandler
+
+import numpy as np
+import pandas as pd
+
+from mindsdb.integrations.handlers.openai_handler.constants import CHAT_MODELS as OPEN_AI_CHAT_MODELS
+from mindsdb.integrations.libs.llm.utils import get_llm_config
+from mindsdb.integrations.utilities.handler_utils import get_api_key
+from mindsdb.integrations.handlers.langchain_embedding_handler.langchain_embedding_handler import construct_model_from_args
+from mindsdb.utilities import log
+from mindsdb.utilities.context_executor import ContextThreadPoolExecutor
+from mindsdb.interfaces.storage import db
+
+from .mindsdb_chat_model import ChatMindsdb
+from .log_callback_handler import LogCallbackHandler
+from .langfuse_callback_handler import LangfuseCallbackHandler
+from .tools import langchain_tools_from_skill
+from .safe_output_parser import SafeOutputParser
+
+from .constants import (
+    DEFAULT_AGENT_TIMEOUT_SECONDS,
+    DEFAULT_AGENT_TYPE,
+    DEFAULT_MAX_ITERATIONS,
+    DEFAULT_MAX_TOKENS,
+    SUPPORTED_PROVIDERS,
+    ANTHROPIC_CHAT_MODELS,
+    OLLAMA_CHAT_MODELS,
+    USER_COLUMN,
+    ASSISTANT_COLUMN
+)
+
+_PARSING_ERROR_PREFIXES = ['An output parsing error occured', 'Could not parse LLM output']
+
+logger = log.getLogger(__name__)
+
+
+def get_llm_provider(args: Dict) -> str:
+    if 'provider' in args:
+        return args['provider']
+    if args['model_name'] in ANTHROPIC_CHAT_MODELS:
+        return 'anthropic'
+    if args['model_name'] in OPEN_AI_CHAT_MODELS:
+        return 'openai'
+    if args['model_name'] in OLLAMA_CHAT_MODELS:
+        return 'ollama'
+    raise ValueError("Invalid model name. Please define provider")
+
+
+def get_embedding_model_provider(args: Dict) -> str:
+    if 'embedding_model_provider' in args:
+        return args['embedding_model_provider']
+    if 'embedding_model_provider' not in args:
+        logger.warning('No embedding model provider specified. trying to use llm provider.')
+        return args.get('embedding_model_provider', get_llm_provider(args))
+    raise ValueError("Invalid model name. Please define provider")
+
+
+def get_chat_model_params(args: Dict) -> Dict:
+    model_config = args.copy()
+    # Include API keys.
+    model_config['api_keys'] = {
+        p: get_api_key(p, model_config, None, strict=False) for p in SUPPORTED_PROVIDERS
+    }
+    llm_config = get_llm_config(args.get('provider', get_llm_provider(args)), model_config)
+    config_dict = llm_config.model_dump()
+    config_dict = {k: v for k, v in config_dict.items() if v is not None}
+    return config_dict
+
+
+def create_chat_model(args: Dict):
```

Comment:
I don't want to move it somewhere to libs/llm/utils because it adds dependency from langchain there (and also from ChatMindsdb)
I can create new file next to `langchain_agent` file if it works.